File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter17/FingerPaint/FingerPaint/AppSettings.cpp
#include "pch.h"
#include "AppSettings.h"

using namespace FingerPaint;

using namespace Platform;
using namespace Windows::Foundation::Collections;
using namespace Windows::Storage;
using namespace Windows::UI;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Media;

AppSettings::AppSettings()
{
    IPropertySet^ props = ApplicationData::Current->LocalSettings->Values;
    this->LoadedFilePath = props->HasKey("LoadedFilePath") ? dynamic_cast<String^>(props->Lookup("LoadedFilePath")) : nullptr;
    this->LoadedFilename = props->HasKey("LoadedFilename") ? dynamic_cast<String^>(props->Lookup("LoadedFilename")) : nullptr;
    this->isImageModified = props->HasKey("IsImageModified") ? (bool)props->Lookup("IsImageModified") : false;
    this->Thickness = props->HasKey("Thickness") ? (float)props->Lookup("Thickness") : 16;

    if (props->HasKey("Color.Red") && props->HasKey("Color.Green") && props->HasKey("Color.Blue"))
    {
        this->Color = ColorHelper::FromArgb(255,
                                            (byte)props->Lookup("Color.Red"),
                                            (byte)props->Lookup("Color.Green"),
                                            (byte)props->Lookup("Color.Blue"));    
    }
    else
    {
        this->Color = Colors::Blue;
    }
}

void AppSettings::Save()
{
    IPropertySet^ props = ApplicationData::Current->LocalSettings->Values;
    props->Clear();
    props->Insert("LoadedFilePath", this->LoadedFilePath);
    props->Insert("LoadedFilename", this->LoadedFilename);
    props->Insert("IsImageModified", this->IsImageModified);
    props->Insert("Thickness", this->Thickness);
    props->Insert("Color.Red", this->Color.R);
    props->Insert("Color.Green", this->Color.G);
    props->Insert("Color.Blue", this->Color.B);
}

void AppSettings::LoadedFilePath::set(String^ value)
{
    if (!loadedFilePath->Equals(value))
    {
        loadedFilePath = value;
        OnPropertyChanged("LoadedFilePath");
    }
}

void AppSettings::LoadedFilename::set(String^ value)
{
    if (!loadedFilename->Equals(value))
    {
        loadedFilename = value;
        OnPropertyChanged("LoadedFilename");
    }
}

void AppSettings::IsImageModified::set(bool value)
{
    if (isImageModified != value)
    {
        isImageModified = value;
        OnPropertyChanged("IsImageModified");
    }
}

void AppSettings::Thickness::set(float value)
{
    if (thickness != value)
    {
        thickness = value;
        OnPropertyChanged("Thickness");
    }
}

void AppSettings::Color::set(Windows::UI::Color value)
{
    if (!color.Equals(value))
    {
        color = value;
        OnPropertyChanged("Color");
    }
}

void AppSettings::OnPropertyChanged(String^ propertyName)
{
    PropertyChanged(this, ref new PropertyChangedEventArgs(propertyName));
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter14/FingerPaint/FingerPaint/AppSettings.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using Windows.Storage;
using Windows.UI;

namespace FingerPaint
{
    public class AppSettings : INotifyPropertyChanged
    {
        // Application settings initial values
        string loadedFilePath = null;
        string loadedFilename = null;
        bool isImageModified = false;
        Color color = Colors.Blue;
        double thickness = 16;

        public event PropertyChangedEventHandler PropertyChanged;

        public AppSettings()
        {
            ApplicationDataContainer appData = ApplicationData.Current.LocalSettings;

            if (appData.Values.ContainsKey("LoadedFilePath"))
                this.LoadedFilePath = (string)appData.Values["LoadedFilePath"];

            if (appData.Values.ContainsKey("LoadedFilename"))
                this.LoadedFilename = (string)appData.Values["LoadedFilename"];

            if (appData.Values.ContainsKey("IsImageModified"))
                this.IsImageModified = (bool)appData.Values["IsImageModified"];

            if (appData.Values.ContainsKey("Color.Red") &&
                appData.Values.ContainsKey("Color.Green") &&
                appData.Values.ContainsKey("Color.Blue"))
            {
                this.Color = Color.FromArgb(255,
                                            (byte)appData.Values["Color.Red"],
                                            (byte)appData.Values["Color.Green"],
                                            (byte)appData.Values["Color.Blue"]);
            }

            if (appData.Values.ContainsKey("Thickness"))
                this.Thickness = (double)appData.Values["Thickness"];

        }

        public string LoadedFilePath
        {
            set { SetProperty<string>(ref loadedFilePath, value); }
            get { return loadedFilePath; }
        }

        public string LoadedFilename
        {
            set { SetProperty<string>(ref loadedFilename, value); }
            get { return loadedFilename; }
        }

        public bool IsImageModified
        {
            set { SetProperty<bool>(ref isImageModified, value); }
            get { return isImageModified; }
        }

        public Color Color
        {
            set { SetProperty<Color>(ref color, value); }
            get { return color; }
        }

        public double Thickness
        {
            set { SetProperty<double>(ref thickness, value); }
            get { return thickness; }
        }

        public void Save()
        {
            ApplicationDataContainer appData = ApplicationData.Current.LocalSettings;
            appData.Values.Clear();
            appData.Values.Add("LoadedFilePath", this.LoadedFilePath);
            appData.Values.Add("LoadedFilename", this.LoadedFilename);
            appData.Values.Add("IsImageModified", this.IsImageModified);
            appData.Values.Add("Color.Red", this.Color.R);
            appData.Values.Add("Color.Green", this.Color.G);
            appData.Values.Add("Color.Blue", this.Color.B);
            appData.Values.Add("Thickness", this.Thickness);
        }

        protected bool SetProperty<T>(ref T storage, T value,
                                      [CallerMemberName] string propertyName = null)
        {
            if (object.Equals(storage, value))
                return false;

            storage = value;
            OnPropertyChanged(propertyName);
            return true;
        }

        protected void OnPropertyChanged(string propertyName)
        {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}
